### PR TITLE
[wasm] change filtering system timezones from zone.tab as a task parameter

### PR DESCRIPTION
**Pull request by tqiu8**
Partial fix for #44840

Default number of system timezones is now 349. There is no impact on `dotnet.timezones.blat` size.

### CODE REVIEW

File: tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs [hunk 1]
```diff
@@ -81,17 +84,21 @@ public override bool Execute()
             Log.LogError($"Could not find required file {ZoneTabFile}");
             return false;
         }
-
         CompileTimeZoneDataSource();
 
-        string[] filtered = new string[] { "America/Los_Angeles", "Australia/Sydney", "Europe/London", "Pacific/Tongatapu",
+        if (FilterSystemTimeZones!) {
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs:89]: Why is the `!` needed here? Also, opening brace should be on a new line.

Comment (radical) [reply tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs:89]: The `!` is needed when you have a nullable type, like `public string[]? TimeZones`, and you want to tell the compiler that you know it isn't null, so it shouldn't complain about a possible null dereference. And because `FilterSystemTimeZones` is not nullable, `!` isn't required here. (https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving)

File: tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs [hunk 2]
```diff
@@ -81,17 +84,21 @@ public override bool Execute()
...
+            string[] filtered = new string[] { "America/Los_Angeles", "Australia/Sydney", "Europe/London", "Pacific/Tongatapu",
                                 "America/Sao_Paulo", "Australia/Perth", "Africa/Nairobi", "Europe/Berlin",
                                 "Europe/Moscow", "Africa/Tripoli", "America/Argentina/Catamarca", "Europe/Lisbon",
                                 "America/St_Johns"};
+            FilterZoneTab(filtered, ZoneTabFile);
+        }
+        else
+        {
+            File.Copy(ZoneTabFile, Path.Combine(OutputDirectory!, "zone.tab"), true);
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs:98]: Is this *new* intended behavior, to have `zone.tab` in the output directory, as the output of the task?

File: tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs [hunk 3]
```diff
@@ -49,7 +51,8 @@ private void FilterTimeZoneData()
         //  for ex: `CST6CDT`, `MST`, etc.
         foreach (var entry in new DirectoryInfo (OutputDirectory!).EnumerateFiles())
         {
-            File.Delete(entry.FullName);
+            if (entry.Name != "zone.tab")
```
Comment (radical) [inline tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs:54]: Why is this needed? We never copy this file to the `OutputDirectory`, so it shouldn't be there at all, IIUC.

File: tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs (hunk 1 above)
Comment (tqiu8) [reply tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs:89]: Oops it was original nullable but I forgot to take the exclamation point out.

File: tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs (hunk 2 above)
Comment (tqiu8) [reply tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs:98]: Yes it is intended to have that file in the output directory so that it is bundled into `dotnet.timezones.blat`

File: tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs (hunk 3 above)
Comment (tqiu8) [reply tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs:54]: It is copied to the output directory (read above).

File: src/mono/wasm/wasm.proj [hunk 4]
```diff
@@ -53,6 +53,10 @@
   <UsingTask TaskName="CompileTimeZoneData"
     AssemblyFile="$(WasmBuildTasksAssemblyPath)"/>
   <Target Name="CompileTimeZones">
+    <PropertyGroup>
+      <FilterSystemTimeZones>false</FilterSystemTimeZones>
+    </PropertyGroup>
+
```
Comment (radical) [inline src/mono/wasm/wasm.proj:59]: Move this outside the target. Having this here means that it will always override the value set elsewhere.
Also, change this to: `<FilterSystemTimeZones Condition="'$(FilterSystemTimeZones)' == ''">false</FilterSystemTimeZones>`

File: tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs (hunk 3 above)
Comment (radical) [reply tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs:54]: AFAIU, earlier behavior was:

1. run `zic` # this does not create the zone.tab file
2. `FilterTimeZoneData` deletes bunch of files like for `MST` (does *not* exclude `zone.tab`, but it isn't there, so it doesn't matter)
3. `FilterZoneData` - creates a new `zone.tab` by filtering `zone1970.tab`

Now:

1. run `zic` # this does not create the zone.tab file
2. `FilterZoneData` - creates a new `zone.tab` by filtering `zone1970.tab`
3. `FilterTimeZoneData` deletes bunch of files like for `MST`

So, step 3 creates a `zone.tab`. And because the patch moves step 3 => 2, the condition for skipping `zone.tab` is required. Instead, you can retain the order as it was earlier, and no check is required.

Comment (radical) [reply tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs:54]: I added some logging to the task, which should be useful for debugging, and should make the output cleaner.
https://gist.github.com/radical/00989f366f3f4bc263f06d87dd36c4e1

Comment (tqiu8) [reply tools-local/tasks/mobile.tasks/WasmBuildTasks/CompileTimeZoneData.cs:54]: I looked it over and it looks good to me. Feel free to push to my branch.

**Comment by radical** [comment]
Windows failure seems unrelated:
```
C:\h\w\B07E099F\w\ADEC0929\e>xunit.console.exe System.Security.Cryptography.Pkcs.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing  
  Discovering: System.Security.Cryptography.Pkcs.Tests (app domain = on [no shadow copy], method display = ClassAndMethod, method display options = None)
  Discovered:  System.Security.Cryptography.Pkcs.Tests (found 311 of 380 test cases)
  Starting:    System.Security.Cryptography.Pkcs.Tests (parallel test collections = on, max threads = 2)
    System.Security.Cryptography.Pkcs.Tests.Pkcs9AttributeTests.DocumentNameMissingTerminator [FAIL]
      Assert.Equal() Failure
                         (pos 8)
      Expected: My Name!
      Actual:   My Name!ns
                         (pos 8)
      Stack Trace:
        /_/src/libraries/System.Security.Cryptography.Pkcs/tests/Pkcs9AttributeTests.cs(260,0): at System.Security.Cryptography.Pkcs.Tests.Pkcs9AttributeTests.DocumentNameMissingTerminator()
    System.Security.Cryptography.Pkcs.EnvelopedCmsTests.Tests.DecryptTestsUsingCertWithPrivateKey.DecryptEnvelopedOctetStringWithIndefiniteLength [SKIP]
      Condition(s) not met: "SupportsIndefiniteLengthEncoding"
    System.Security.Cryptography.Pkcs.EnvelopedCmsTests.Tests.DecryptTestsUsingCertWithPrivateKey.DecryptEnvelopedEmptyOctetStringWithIndefiniteLength [SKIP]
      Condition(s) not met: "SupportsIndefiniteLengthEncoding"
    System.Security.Cryptography.Pkcs.EnvelopedCmsTests.Tests.KeyAgreeRecipientInfoTests.TestKeyAgreement_PlatformNotSupported [SKIP]
      Condition(s) not met: "DoesNotSupportDiffieHellman"
  Finished:    System.Security.Cryptography.Pkcs.Tests
=== TEST EXECUTION SUMMARY ===
   System.Security.Cryptography.Pkcs.Tests  Total: 353, Errors: 0, Failed: 1, Skipped: 3, Time: 1.742s
----- end Tue 11/24/2020 19:41:00.26 ----- exit code 1 ----------------------------------------------------------
```

**Review by radical — APPROVED** [review]
LGTM, as far as making timezone filtering optional 👍
